refactor the indexer

Omarabdul3ziz opened the pull request:
## Description & Changes
- combine indexers to be a single indexer with different watchers
- use a single RpcRmbClient across the proxy
- add more flags `-no-indexer` and `no-modify` to the commands & update the workflow to stop the indexer while integration tests
- add watchers for dmi/speed
- implement Scanner/Valuer for model structs
- add triggers on the new table to update the resources cache
- use `gorm` on the crafter for easier mock data generations for the custom model types
- add loader/tests for the new fields

## Related issues
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/664
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/636

## Checklist

- [x] Tests included
- [x] Build pass
- [x] Documentation
- [x] Code format and docstring

Omarabdul3ziz committed:
convert indexers to watchers in a mono indexer

* seperate commit for easier review *

Omarabdul3ziz committed:
use a single rmb client of type peer.RpcClient across the project

Omarabdul3ziz committed:
configure starting/stoping indexer/modifier from flag

Omarabdul3ziz committed:
wip: introducing the dmi watcher

Omarabdul3ziz committed:
upsert dmi on database

Omarabdul3ziz committed:
add network speed watcher

Omarabdul3ziz committed:
implement an Scanner/Valuer interfaces for custom gorm jsonb types in dmi model

Omarabdul3ziz committed:
fix network speed types

Omarabdul3ziz committed:
add triggers for dmi/speed tables

Omarabdul3ziz committed:
add generators:
- organize the schema file & add new indexer tables
- add generator for the new indexer tables
- use gorm instead of plain sql in the crafter

Omarabdul3ziz committed:
add loader/tests for the new dmi/speed data

Omarabdul3ziz committed:
make the interval/worker configurable

Omarabdul3ziz committed:
Merge branch 'development' of https://github.com/threefoldtech/tfgrid-sdk-go into development_proxy_indexer

Omarabdul3ziz committed:
stop data modification, remove unnecessary conversion, silent the gorm on tests

xmonader commented on grid-proxy/cmds/proxy_server/main.go:
```diff
@@ -57,9 +58,13 @@ type flags struct {
-	healthIndexerInterval uint
+	healthIndexerWorkers        uint
+	healthIndexerInterval       uint
+	dmiWatcherWorkers           uint
```
why did we move from the word "indexer" to watcher?

xmonader commented on grid-proxy/pkg/types/nodes.go:
```diff
@@ -53,6 +55,11 @@ type CapacityResult struct {
 	Used  Capacity `json:"used_resources"`
 }
 
+type Speed struct {
```
I noticed there're multiple speeds for upload / download there's ipv4, ipv6 versions? which is it? and are

Omarabdul3ziz replied:
yes we have `ipv4` and `ipv6` for both `tcp` and `udp` these are four per an iperf server, 
for example devnet has two iperf servers on the two public nodes so we have total of `8` tests

a discussion was raised on the chat group and i remember the decision was to show the only non-zero value. should we do something different? or view them all?

xmonader commented on grid-proxy/internal/indexer/utils.go:
```diff
@@ -0,0 +1,34 @@
+package indexer
+
```
rename the file to filter_up_nodes.go

Omarabdul3ziz replied:
it was planned to have more utils method for filtering up nodes/healthy nodes or call rmb

xmonader commented on grid-proxy/internal/indexer/types.go:
```diff
@@ -0,0 +1,47 @@
+//		item2
+//		...
+//	]
+type DMI struct {
```
Why didn't you bring the dmidecoder from zos ?

xmonader commented on grid-proxy/internal/indexer/indexer.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+const (
+	indexerCallTimeout = 10 * time.Second
```
bump that to 30

xmonader commented on grid-proxy/internal/indexer/indexer.go:
```diff
@@ -0,0 +1,54 @@
+	indexerCallTimeout = 10 * time.Second
+)
+
+type Watcher interface {
```
Rename to Indexer

xmonader commented on grid-proxy/internal/indexer/indexer.go:
```diff
@@ -0,0 +1,54 @@
+	Start(ctx context.Context)
+}
+
+type Indexer struct {
```
Rename to Manager

xmonader commented on grid-proxy/internal/indexer/indexer.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+type Watcher interface {
+	Start(ctx context.Context)
```
Should be Starter and Stopper and Pauser interfaces instead of Indexer or Watcher here

xmonader commented on grid-proxy/internal/indexer/indexer.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func (i *Indexer) Start() {
+	if i.Paused {
```
It makes sense to resume after pausing, not sure why you return

Omarabdul3ziz replied:
actually `pause` was a misleading i need a flag to stop the indexer in this run, it is helpful when running integration tests to keep data unchanged from the indexer.

for a better naming i called the flag `-no-indexer` also made the decision for starting manager or not outside the pkg in the main file

xmonader commented on grid-proxy/internal/indexer/indexer.go:
```diff
@@ -0,0 +1,54 @@
+	}
+}
+
+func (i *Indexer) RegisterWatcher(name string, watcher Watcher) {
```
RegisterIndexer or Register is enough
the 2nd param should be: indexer

xmonader commented on grid-proxy/internal/indexer/indexer.go:
```diff
@@ -0,0 +1,54 @@
+	RmbClient *peer.RpcClient
+}
+
+func NewIndexer(
```
NewManager

xmonader commented on grid-proxy/internal/indexer/indexer.go:
```diff
@@ -0,0 +1,54 @@
+
+func NewIndexer(
+	ctx context.Context,
+	paused bool,
```
Shouldn't be here, if for internal state, it should be defaulted to the zero value, false

xmonader commented on grid-proxy/internal/indexer/gpu.go:
```diff
@@ -1,19 +1,13 @@
-	rmbTypes "github.com/threefoldtech/tfgrid-sdk-go/rmb-sdk-go/peer/types"
 )
 
 const (
```
prefix all of the consts with gpu (given there're multiple indexers in the same package)

xmonader commented on grid-proxy/internal/indexer/gpu.go:
```diff
@@ -37,16 +31,15 @@ type NodeGPUIndexer struct {
+	nodesGPUBufferWorkers uint) *NodeGPUIndexer {
+	return &NodeGPUIndexer{
 		db:                     db,
+		rpcClient:              rpcClient,
```
is that client a reconnecting one, in long running processes you should be careful not to lose that single connection you have.

Omarabdul3ziz replied:
afaik yes, the rpcClient taking a substrate manager and handle the open/close connections when needed

xmonader commented on grid-proxy/internal/indexer/gpu.go:
```diff
@@ -130,11 +108,25 @@ func (n *NodeGPUIndexer) runQueryGridNodes(ctx context.Context) {
 
 func (n *NodeGPUIndexer) getNodeGPUInfo(ctx context.Context, nodeTwinID uint32) error {
-	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	subCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
```
bump to 30

xmonader commented on grid-proxy/internal/indexer/dmi.go:
```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+func parseDmiResponse(dmiResponse DMI) types.DmiInfo {
```
replace everything related to parsing by zos dmidecode package, don't do it yourself

Omarabdul3ziz replied:
checked the dmi pkg in zos and the `DMI` response is already parsed from the pkg in zos.
what i am doing here is just selecting what we want to store instead of storing the whole dmi response

xmonader reviewed (changes requested):
I had to stop the review. 

1- Please make sure all of the indexers are having the similar interface 
2- Remove everything related to any manual parsing of the dmidecode
3- Prefix all of the constants related to each indexer with the indexer name e.g `healthIndexerNumWorkers`
4- for every indexer there should be
 - a `Start` method, can have its own interface `Starter`
 - a `Stop` method, can have its own interface `Stopper`
 - a `FindHealthyNodes` method
 - a `ResultsBatcher` method
 - a `DBBatchesUpserter` method
 
 All of them can be converted to functions whenever needed.
 
 Each indexer file should have the logic top to bottom 
 
 e.g 
 
- consts
- vars
- types
- factories
- entry point `Start` function
- then each of the methods asked for 

Note: consts and vars should be prefixed with the indexer name

Use the same terms everywhere: e.g status, state, numWorkers, workersCount, .. etc All of these need to be unified


in an upcoming improvement we can use singleflight to avoid expensive DB/network calls e.g all are asking for the healthy nodes within the same timeframe

Aside from that, the PR already has too many concerns, supporting the dmi, refactoring in the indexers, introducing more dependencies. Please in the future make sure it's scoped.

Omarabdul3ziz committed:
Merge branch 'development' of https://github.com/threefoldtech/tfgrid-sdk-go into development_proxy_indexer

Omarabdul3ziz committed:
refactor the indexer manager to implement a unified interface for indexers:

- configure a flag to stop or start the indexer manager on the startup. it is useful to use `-no-indexer` flag while running integration tests, to keep the loaded data in the mock client consistent with the actual data in DB.
- organize the database calls in `postgres.go` file by separating the indexer calls into a new file. also, save consistency for numeric `NodeTwinId` filed to be `uint32` in all methods
- remove the duplicated types across different pkgs and use only the ones in `pkg/types` pkgs db/tests/generator
- add new trigger on the `node_gpu` table, this trigger should be fired on insert/delete to update the `num_gpu` filed on the `resources_cache` table
- remove all implementation for `Scanner/Valuer` interfaces for jsonb fields in gorm model. instead easier and cleaner, use the registered json serializer. apply this for dmi fields/node power object
- modify the generation of `healthy_reports` mock data to use `gorm` functionality instead of creating a plain SQL string
- update and organize the schema file used on the db mock generation to be aligned with the new defined struct models.

Indexer pkg changes:

- import types directly from zos once needed instead of re-implement it, applied this for IperfTestResult and DMI
- create utils method, generic methods that can be used in all indexers to avoid repeating the code
- create a manager that register/control all the indexers
- define an `Indexer` interface that defines the needed methods for each indexer
- document the indexer package along with the registered indexers and their metadata
- implement `Indexer` interface for each registered interface, this requires some refactoring for the indexer especially the GPU indexer
- a better decision on old GPU node deletion by adding a timestamp expiration and check it is expired or not with each upsert, the allowed interval is the same as the check interval in the indexer an hour for now. this is better than checking the lastAdded twin id cause we have multiple caller/batcher that may conflict the results.
- add new trigger for the dmi indexer, to start querying the newly added nodes and don’t wait for the new rotation

Omarabdul3ziz committed:
Merge branch 'development' of https://github.com/threefoldtech/tfgrid-sdk-go into development_proxy_indexer

Omarabdul3ziz committed:
clean & complete some missing parts:
- correctly handle the indexer manager start is !noIndexer after debugging
- change the updated_at field on gpu table in schema to be a bigint to align with the struct model
- use gorm instead of plain sql for scanning in loaders in both load nodes/dmi to benefit from the registered json serializer instead of implementing the scanning logic for the jsonb fields

Omarabdul3ziz commented:
Update changes:
- configure a flag to stop or start the indexer manager on the startup. it is useful to use `-no-indexer` flag while running integration tests, to keep the loaded data in the mock client consistent with the actual data in DB.
- organize the database calls in `postgres.go` file by separating the indexer calls into a new file. also, save consistency for numeric `NodeTwinId` filed to be `uint32` in all methods
- remove the duplicated types across different pkgs and use only the ones in `pkg/types` pkgs db/tests/generator
- add new trigger on the `node_gpu` table, this trigger should be fired on insert/delete to update the `num_gpu` filed on the `resources_cache` table
- remove all implementation for `Scanner/Valuer` interfaces for jsonb fields in gorm model. instead easier and cleaner, use the registered json serializer. apply this for dmi fields/node power object
- modify the generation of `healthy_reports` mock data to use `gorm` functionality instead of creating a plain SQL string
- update and organize the schema file used on the db mock generation to be aligned with the new defined struct models.

Indexer pkg changes:
- import types directly from zos once needed instead of re-implement it, applied this for IperfTestResult and DMI
- create utils method, generic methods that can be used in all indexers to avoid repeating the code
- create a manager that register/control all the indexers
- define an `Indexer` interface that defines the needed methods for each indexer
- document the indexer package along with the registered indexers and their metadata
- implement `Indexer` interface for each registered interface, this requires some refactoring for the indexer especially the GPU indexer
- a better decision on old GPU node deletion by adding a timestamp expiration and check it is expired or not with each upsert, the allowed interval is the same as the check interval in the indexer an hour for now. this is better than checking the lastAdded twin id cause we have multiple caller/batcher that may conflict the results.
- add new trigger for the dmi indexer, to start querying the newly added nodes and don’t wait for the new rotation

Omarabdul3ziz committed:
update mod files

Omarabdul3ziz committed:
Merge branch 'development' of https://github.com/threefoldtech/tfgrid-sdk-go into development_proxy_indexer

Omarabdul3ziz committed:
silent the gorm logs in dump data generation

muhamadazmy commented on grid-proxy/cmds/proxy_server/main.go:
```diff
@@ -81,13 +86,18 @@ func main() {
+	flag.UintVar(&f.dmiIndexerIntervalMins, "dmi-indexer-interval", 60*24, "node dmi check interval in min")
+	flag.UintVar(&f.dmiIndexerNumWorkers, "dmi-indexer-workers", 1, "number of workers checking on node dmi")
+	flag.UintVar(&f.speedIndexerIntervalMins, "speed-indexer-interval", 5, "node speed check interval in min")
+	flag.UintVar(&f.speedIndexerNumWorkers, "speed-indexer-workers", 100, "number of workers checking on node speed")
```
Just wondering why do we need multiple indexer workers to query things from nodes, can't we have a single set of workers that they ask the node for the same thing for everything? 

I mean i see here, dmi-indexer-workers, gpu-indexer-workers, speed-indexer-workers (whatever that means), etc... 

This is just a question, not request for change

muhamadazmy commented on grid-proxy/internal/indexer/gpu.go:
```diff
@@ -0,0 +1,177 @@
+
+func (n *NodeGPUIndexer) Start(ctx context.Context) {
+	go n.StartNodeFinder(ctx)
+	go n.startNodeTableWatcher(ctx)
```
Do these methods need to be public? I believe they are intended for private use by the indexer object. I suggest they are renamed to be private instaed

muhamadazmy replied:
This applies to all other indexers here

muhamadazmy commented on grid-proxy/internal/indexer/health.go:
```diff
@@ -0,0 +1,139 @@
+	batchSize       uint
+}
+
+func NewNodeHealthIndexer(
```
I can see that the an indexer mainly does the same operation, I was wondering that an indexer logic can be abstracted more. 
From what i see an indexer has a set of options (say workers), he run indexing every certain amount of time. Then does some sort of batching to update the db. 
All this logic then can be unified. Then each indexer need to define 2 things.
- The function that does the "fetch" of the data. 
- The logic of "batch" db update.

Those 2 functions then can be part of an IndexerWork interface. And can has multiple implementation (say gpu and dmi) and then an Indexer instance can be created for each something like `NewIndexer(GPUIndexerWork())` for example. Then call `indexer.Start()`

Omarabdul3ziz committed:
abstract the indexer pkg code using generics

Description:

- redefining the indexer structs to be a single generic indexer without manager, each indexer should do the generic functions of find/get/batch/upsert results. with each indexer have different job based on the defined type will creating the indexer.
- each indexer job should only handle its specific job of:
    - just defining its finders or triggers
    - preparing the rmb call to the node, parse the response and return a ready db model template that can stored
    - upserting its restults by updating or descarding or inserting the passed arguments
- the indexer code is the one responsible for doing the logic, by
    - executing the finders methods for each indexer as it is defined in its job based on the interval
    - starting the getter worker that invoke the get method on the work
    - starting a batcher for each indexer, that have two channels as its ends
    - starting the upserter worker that invoke the upsert method on the work

Changes:

- remove the indexer manager for easily starting a separate generic indexer based on the resource type.
- remove the redundant `NewIndexer` `Start` `Find` `Batch` methods for each indexer for a generic method on the `Indexer[T any]` struct
- update the return of the get method on work to be `[]T` instead `T` to cover all the types indexer responses like gpu call which return a slice of gpus
- creating the finders map for easly reusablity of the finders method on new/up/health nodes
- check the uniqueness of the nodes on the batch before inserting to prevent having duplicated data comming from two different triggers like healthy/up or up/new
- modify the default sql-log-level in the makefile
- add more debug logging statement
- update docs

Omarabdul3ziz commented:
Converting to draft to not merge till 3.13 reaches mainnet.

Omarabdul3ziz committed:
Merge branch 'development' of https://github.com/threefoldtech/tfgrid-sdk-go into development_proxy_indexer

Omarabdul3ziz committed:
remove already merged trigger on node_gpu

xmonader commented:
yalla beena
